group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java

```diff
@@ -0,0 +1,998 @@
+
+            throw new IllegalStateException("None of the member's protocols can be supported.");
+        }
+
+        if (!leaderId.isPresent()) {
+            leaderId = Optional.of(member.memberId());
+        }
+
+        members.put(member.memberId(), member);
+        incrementSupportedProtocols(member);
+        member.setAwaitingJoinFuture(future);
+
+        if (member.isAwaitingJoin()) {
+            numMembersAwaitingJoinResponse++;
+        }
+
+        pendingJoinMembers.remove(member.memberId());
+    }
+
+    /**
+     * Remove a member from the group.
+     *
+     * @param memberId the member id to remove.
+     */
+    public void remove(String memberId) {
+        GenericGroupMember removedMember = members.remove(memberId);
+        if (removedMember != null) {
+            decrementSupportedProtocols(removedMember);
+            if (removedMember.isAwaitingJoin()) {
+                numMembersAwaitingJoinResponse--;
+            }
+
+            removedMember.groupInstanceId().ifPresent(staticMembers::remove);
+        }
+
+        if (isLeader(memberId)) {
+            Iterator<String> iter = members.keySet().iterator();
+            String newLeader = iter.hasNext() ? iter.next() : null;
+            leaderId = Optional.ofNullable(newLeader);
+        }
+
+        pendingJoinMembers.remove(memberId);
+        pendingSyncMembers.remove(memberId);
+    }
+
+    /**
+     * Check whether current leader has rejoined. If not, choose a
+     * new leader from one of the joined members.
+     *
+     * Return false if
+     *   1. the group is currently empty (has no designated leader)
+     *   2. no member rejoined
+     *
+     * @return whether a new leader was elected.
+     */
+    public boolean maybeElectNewJoinedLeader() {
+        if (leaderId.isPresent()) {
+            GenericGroupMember currentLeader = member(leaderId.get());
+            if (!currentLeader.isAwaitingJoin()) {
+                for (GenericGroupMember member : members.values()) {
+                    if (member.isAwaitingJoin()) {
+                        leaderId = Optional.of(member.memberId());
+                        log.info("Group leader [memberId: {}, groupInstanceId: {}] " +
+                            "failed to join before the rebalance timeout. Member {} " +
+                            "was elected as the new leader.",
+                            currentLeader.memberId(),
+                            currentLeader.groupInstanceId().orElse("None"),
+                            member
+                        );
+                        return true;
+                    }
+                }
+                log.info("Group leader [memberId: {}, groupInstanceId: {}] " +
+                        "failed to join before the rebalance timeout and the " +
+                        "group couldn't proceed to the next generation because " +
+                        "no member joined.",
```

Comment:
nit: The indentation of these lines it not consistent with the previous `log.info` at L409. Let's try to use the same format.